New TechnologyHelper and TechBoard integration for Strategy Card Automator

jirach08 opened the pull request:
Created a TechnologyHelper with list of all tech (including faction tech), type, requirements, abbreviation and expansion they belong to along with getters.

Fixed a typo in FactionHelper

added onNewTechnologyResearched event to UnitHighlightsUpdater to better identify the artificial event

updated StrategyCardAutomator
 - improved description text
 - removed old comments
 - updated Technology event to account for Technology Board and no longer stack multiple Technology Cards next to the deck

jirach08 committed:
Revert "Revert "added TI4 Technology Helper""

This reverts commit 788c12becb847551f2692582449193a0f24d158c.

jirach08 committed:
new event broadcast to all listeners onNewTechnologyResearched

jirach08 committed:
Merge branch 'master' into techHelper

jirach08 committed:
techBoard integration WIP

jirach08 committed:
Added a queue to choose technology cards
Unit Upgrades are now copied to faction sheet if the tech board is out.

jirach08 committed:
Merge branch 'master' of https://github.com/darrellanderson/CrLua

jirach08 committed:
Merge branch 'master' into techHelper

jirach08 committed:
- add function to match card name to list of known technologies in TechnologyHelper
- remove experimental technology Queue, instead make takeObject smooth=false so that they appear "instantly"
	- this is to prevent subsequent calls of researchTech from stacking tech cards

jirach08 committed:
adjusted the sensitivity when avoiding overlapping cards with researchTech
removed remnants of queueResearchTech test
control for .." \Omega" names in technology
fixed a typo in faction helper

 On branch techHelper
 Changes to be committed:
	modified:   Objects/StrategyCardAutomator.ttslua
	modified:   Objects/TI4_Helpers/TI4_FactionHelper.ttslua
	modified:   Objects/TI4_Helpers/TI4_TechnologyHelper.ttslua

jirach08 committed:
cleaned up comments and improved description

darrellanderson commented on Objects/StrategyCardAutomator.ttslua:
```diff
@@ -1623,10 +1624,20 @@ function researchTech(color, cardName)
+	local techCardPositionsOnTable = {}
+	local techType = _technologyHelper.getTechType(cardName)
+	for _,object in ipairs(getAllObjects()) do
+		if (color == _zoneHelper.zoneFromPosition(object.getPosition())) then
```
I don't know how expensive cross-objects calls are; for this kind of thing I like to:
1. get the set of tech card names in a single call (e.g. _technologyHelper.techNameSet()).
2. in the loop test the name against  build up guidToPosition for tech.
3. do the zone check for boards and decks only after the name matches so there would only be ~12 in a 6 player game.

Then after the table scan can do the bulk zonesFromPositions on the tech cards, finally extracting just the ones in zone.

jirach08 replied:
I didn't realize, that cross-object calls might be an issue. In that case this makes perfect sense.

darrellanderson commented on Objects/StrategyCardAutomator.ttslua:
```diff
@@ -1637,108 +1648,172 @@ function researchTech(color, cardName)
+		column = _columns[_technologyHelper.getTechType(cardName)]
+		pos.x = pos.x - column*dx
+
+		-- find first spot not occupied by tech card
```
This loop structure could probably be more efficient, e.g. build the set of taken points in the column in a single pass, then find the first available from that smaller set.  This is probably ok as you've already restricted to just tech cards in this player's area.

darrellanderson commented on Objects/StrategyCardAutomator.ttslua:
```diff
@@ -1637,108 +1648,172 @@ function researchTech(color, cardName)
 			pos.x = pos.x - 4
 		end
+
+		-- check that the card is not moved over another existing tech card
```
Could move this nearly identical "find an open spot" into a shared function that takes p0, dx, dz to find the first position = p0 + i * (dx, dz).

jirach08 replied:
Made a getFirstVacantPosition() in the technologyHelper. Might make more sense elsewhere, as this can potentially be a collision check for any object.

darrellanderson commented on Objects/TI4_Helpers/TI4_TechnologyHelper.ttslua:
```diff
@@ -0,0 +1,174 @@
+end
+
+local _technologies = {
+	["Neural Motivator"] = { type = "Green", requirements = nil, abbrev = 'Neural', expansion="base" },
```
Other helpers use `source="PoK"` rather an `expansion`, switch to source for consistency?

jirach08 replied:
I'll just change it then. Should I keep source="base" for base-game components or just remove the entry to let calls return nil?

darrellanderson replied:
faction helper doesn't specify a source for the base factions, but deck helper uses 'Base'.  Game setup is the only consumer right now, and it only looks for "PoK" or "Codex 1" so at the moment it doesn't matter.

darrellanderson commented on Objects/TI4_Helpers/TI4_TechnologyHelper.ttslua:
```diff
@@ -0,0 +1,174 @@
+	["Supercharge"] = { type = "Red", requirements = "R", abbrev = 'Supercharge', expansion = "base" },
+	["Pre-Fab Arcologies"] = { type = "Green", requirements = "GGG", abbrev = 'Pre-Fab Arc', expansion = "base" },
+	-- codex?
+	['????_REDACTED_????'] = { abbrev = 'Scenario Destroyer'},
```
REDACTED is `{ type = "unitUpgrade", requirements = "RR",  abbrev = 'Scenario Destroyer', source = "Codex 1" }`
EXCEPTION is `{ type = "Yellow", requirements = "Y", abbrev = 'Scenario Tech', source = "Codex 1" }`

darrellanderson reviewed (changes requested):
Looks good!  I'm not sure there will be many consumers of the tech helper short term, but having it enables things we couldn't do before.  I'm going to mark this as needing changes before I'll merge it, mainly the tech helper "source" field name.  I'll let you use your discretion for the other comments, they're "would be nice" but not serious problems.

If it's easier I can also accept this update, and do any changes over that with a future update -- just let me know.

darrellanderson commented:
It occurs to me maybe the code to position a technology could move to the helper.  If it exposes `getTechBoardOpenPosition(techCardName)` (or whatever you want to name it) that returned nil/false if no tech board, setup helper could call that to place the starting tech when unpacking a faction.  It would probably use a delay between placements in case there were multiple tech in the same color column.

jirach08 committed:
=
In TechnologyHelper
- Moved getFirstVacantPosition and getVacantPosOnTechBoard to technologyHelper
- added getTechNameSet to get names to all valid technologies
- changed "expansion" entry to "source"
- removed "source" entry for base-game technologies
- added info for "REDACTED" and "EXCEPTION" entries

In StrategyCardAutomator changed researchTechnology():
- reduced number of cross-object calls in getAllObjects()
- outsourced collision checks to technologyHelper

	modified:   Objects/StrategyCardAutomator.ttslua
	modified:   Objects/TI4_Helpers/TI4_TechnologyHelper.ttslua

jirach08 commented:
The getVacantPosOnTechBoard(params) should take care of what you had in mind. It needs the techCardName and the playerColor, and fetches the techBoard and occupiedPositions on its own, if they are not provided. This way, I can also use it for my StrategyCardAutomator without a second getAllObjects() call (by providing techBoard and occupiedPositions).

I also fixed your other suggestions, hopefully to your satisfaction.

jirach08 commented on Objects/StrategyCardAutomator.ttslua:
```diff
@@ -1799,26 +1764,26 @@ function _getEntryGuid(container, namePattern)
+			end
+			Wait.frames(callHandler, i)
+		end
+	end
```
changed spaces to tab indentation.

jirach08 commented on Objects/TI4_Helpers/TI4_TechnologyHelper.ttslua:
```diff
@@ -1,116 +1,118 @@
+		if not helperObject.getVar(functionName) then error('missing ' .. helperObjectName .. '.' .. functionName) end
+		return function(parameters) return helperObject.call(functionName, parameters) end
+	end
+return setmetatable({}, { __index = function(t, k) return getCallWrapper(k) end })
```
changed spaces to tab indentation

darrellanderson reviewed (approved):
Great, thanks for the updates!
